src/Language/Haskell/LSP/Core.hs

```diff
@@ -95,29 +96,37 @@ data ProgressData = ProgressData { progressNextId :: !Int
 
 -- ---------------------------------------------------------------------
 
--- | Language Server Protocol options supported by the given language server.
--- These are automatically turned into capabilities reported to the client
--- during initialization.
+-- | Language Server Protocol options that the server may configure.
+-- If you set handlers for some requests, you may need to set some of these options.
 data Options =
   Options
     { textDocumentSync                 :: Maybe J.TextDocumentSyncOptions
-    , completionProvider               :: Maybe J.CompletionOptions
-    , signatureHelpProvider            :: Maybe J.SignatureHelpOptions
-    , typeDefinitionProvider           :: Maybe J.GotoOptions
-    , implementationProvider           :: Maybe J.GotoOptions
-    , codeActionProvider               :: Maybe J.CodeActionOptions
-    , codeLensProvider                 :: Maybe J.CodeLensOptions
-    , documentOnTypeFormattingProvider :: Maybe J.DocumentOnTypeFormattingOptions
-    , renameProvider                   :: Maybe J.RenameOptions
-    , documentLinkProvider             :: Maybe J.DocumentLinkOptions
-    , colorProvider                    :: Maybe J.ColorOptions
-    , foldingRangeProvider             :: Maybe J.FoldingRangeOptions
-    , executeCommandProvider           :: Maybe J.ExecuteCommandOptions
+    -- |  The characters that trigger completion automatically.
+    , completionTriggerCharacters      :: Maybe [Char]
+    -- | The list of all possible characters that commit a completion. This field can be used
+    -- if clients don't support individual commmit characters per completion item. See
+    -- `_commitCharactersSupport`.
+    , completionAllCommitCharacters    :: Maybe [Char]
+    -- | The characters that trigger signature help automatically.
+    , signatureHelpTriggerCharacters   :: Maybe [Char]
+    -- | List of characters that re-trigger signature help.
+    -- These trigger characters are only active when signature help is already showing. All trigger characters
+    -- are also counted as re-trigger characters.
+    , signatureHelpRetriggerCharacters :: Maybe [Char]
+    -- | CodeActionKinds that this server may return.
+    -- The list of kinds may be generic, such as `CodeActionKind.Refactor`, or the server
+    -- may list out every specific kind they provide.
+    , codeActionKinds                  :: Maybe [J.CodeActionKind]
```

Comment:
I think the API for this isn’t quite right (but that’s not new to this PR):
The LSP spec states

> The `CodeActionOptions` return type is only valid if the client signals code action literal support via the property  `textDocument.codeAction.codeActionLiteralSupport`.

However, the current API requires you to specify the options before you can inspect the capabilities so you can’t actually specify this properly.

Reply:
Good catch. I think we can just return `CodeActionOptionsStatic True` in the case when the client capabilities do not declare support for code action literals, and pass on the code action options in the case that they do